samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -143,9 +230,23 @@ void handleContainerLaunchFail(String processorId, String containerId, String pr
    * @param resourceRequestState state of request in {@link ContainerAllocator}
    */
   @VisibleForTesting
-  void handleExpiredRequestWithHostAffinityEnabled(String processorId, String preferredHost,
+  void handleExpiredRequestForControlActionOrHostAffinityEnabled(String processorId, String preferredHost,
       SamzaResourceRequest request, ContainerAllocator allocator, ResourceRequestState resourceRequestState) {
     boolean resourceAvailableOnAnyHost = allocator.hasAllocatedResource(ResourceRequestState.ANY_HOST);
+
+    if (hasActiveContainerPlacementAction(processorId)) {
+      resourceRequestState.cancelResourceRequest(request);
+      handleControlActionFailure(processorId, getPlacementActionMetadata(processorId).get(),
```

Comment:
The handleControlActionFailure should be called markControlActionFailed, all it does is update the state.